js/modules/timeline.src.js

```diff
@@ -116,90 +127,267 @@ seriesType('timeline', 'line', {
-            ) * multiplier - (padding * 2));
-
-        // Adjust data label width to the currently available value.
-        dataLabel.css({
-            width: dataLabelWidth
-        });
+            userDLOptions = point.userDLOptions || {},
+            multiplier = dataLabelsOptions.alternate ?
+                (isFirstOrLast ? 1.5 : 2) :
+                1,
+            distance,
+            availableSpace = Math.floor(series.xAxis.len / filteredMapLen),
+            pad = dataLabel.padding,
+            targetDLWidth;
+
+        // Adjust data label width to the currently available space.
+        if (point.visible) {
+            distance = Math.abs(userDLOptions.x || point.options.dataLabels.x);
+            if (isInverted) {
+                targetDLWidth = (distance - pad) * 2 - (point.itemHeight / 2);
+
+                dataLabel.css({
+                    width: targetDLWidth,
+                    // Apply ellipsis when data label height is exceeded.
+                    textOverflow: dataLabel.width / targetDLWidth *
+                        dataLabel.height / 2 > availableSpace * multiplier ?
+                            'ellipsis' : 'none'
+                }).shadow({});
+            } else {
+                targetDLWidth = userDLOptions.width ||
+                    dataLabelsOptions.width ||
+                    availableSpace * multiplier - (pad * 2);
+
+                dataLabel.css({
+                    width: targetDLWidth
+                }).shadow({});
+            }
+        }
         Series.prototype.alignDataLabel.apply(series, arguments);
+
+
+    },
+    processData: function () {
+        var series = this;
+
+        series.xData = series.getMappedXData();
+        series.yData = series.xData.map(function (data) {
+            return data === null ? null : 1;
+        });
+
+        Series.prototype.processData.call(this, arguments);
+    },
+    generatePoints: function () {
+        var series = this;
+
+        Series.prototype.generatePoints.apply(series);
+        series.points.forEach(function (point, i) {
+            point.applyOptions({
+                x: series.xData[i]
+            });
+        });
+    },
+    getVisibilityMap: function () {
+        var series = this,
+            map = (series.data.length ?
+                    series.data : series.userOptions.data
+                ).map(function (point) {
+                    return (
+                        point &&
+                        point.visible !== false &&
+                        !point.isNull
+                    ) ? point : false;
+                });
+
+        return map;
     },
     getMappedXData: function () {
         var series = this,
             xMap = [],
-            pointsLen = series.xData.length,
-            max = series.xAxis.options.max,
-            base = max / pointsLen;
+            base,
+            visiblePoints = 0,
+            visibilityMap = series.visibilityMap = series.getVisibilityMap(),
```

Comment:
Do we need to declare `visibilityMap` ? Its user 2 times, so `series.visibilityMap` is enough